packages/ozone/src/daemon/event-pusher.ts

```diff
@@ -0,0 +1,295 @@
+        did: string
+      }
+      pds?: {
+        url: string
+        did: string
+      }
+    },
+  ) {
+    if (services.appview) {
+      this.appview = {
+        agent: new AtpAgent({ service: services.appview.url }),
+        did: services.appview.did,
+      }
+    }
+    if (services.pds) {
+      this.pds = {
+        agent: new AtpAgent({ service: services.pds.url }),
+        did: services.pds.did,
+      }
+    }
+  }
+
+  start() {
+    this.poll(this.repoPollState, () => this.pushRepoEvents())
+    this.poll(this.recordPollState, () => this.pushRecordEvents())
+    this.poll(this.blobPollState, () => this.pushBlobEvents())
+  }
+
+  poll(state: PollState, fn: () => Promise<boolean>) {
+    if (this.destroyed) return
+    state.promise = fn()
+      .then((hadEvts: boolean) => {
+        if (hadEvts) {
+          state.tries = 0
+        } else {
+          state.tries++
+        }
+      })
+      .catch((err) => {
+        dbLogger.error({ err }, 'event push failed')
+        state.tries++
+      })
+      .finally(() => {
+        state.timer = setTimeout(
+          () => this.poll(state, fn),
+          exponentialBackoff(state.tries),
+        )
+      })
+  }
+
+  async processAll() {
+    await Promise.all([
+      this.pushRepoEvents(),
+      this.pushRecordEvents(),
+      this.pushBlobEvents(),
+    ])
+  }
+
+  async destroy() {
+    this.destroyed = true
+    const destroyState = (state: PollState) => {
+      if (state.timer) {
+        clearTimeout(state.timer)
+      }
+      return state.promise
+    }
+    await Promise.all([
+      destroyState(this.repoPollState),
+      destroyState(this.recordPollState),
+      destroyState(this.blobPollState),
+    ])
+  }
+
+  async pushRepoEvents() {
+    return await this.db.transaction(async (dbTxn) => {
+      const toPush = await dbTxn.db
+        .selectFrom('repo_push_event')
+        .selectAll()
+        .forUpdate()
+        .skipLocked()
+        .where('confirmedAt', 'is', null)
+        .where('attempts', '<', 10)
+        .execute()
+      if (toPush.length === 0) return false
+      await Promise.all(toPush.map((evt) => this.attemptRepoEvent(dbTxn, evt)))
+      return true
+    })
+  }
+
+  async pushRecordEvents() {
+    return await this.db.transaction(async (dbTxn) => {
+      const toPush = await dbTxn.db
+        .selectFrom('record_push_event')
+        .selectAll()
+        .forUpdate()
+        .skipLocked()
+        .where('confirmedAt', 'is', null)
+        .execute()
+      if (toPush.length === 0) return false
+      await Promise.all(
+        toPush.map((evt) => this.attemptRecordEvent(dbTxn, evt)),
+      )
+      return true
+    })
+  }
+
+  async pushBlobEvents() {
+    return await this.db.transaction(async (dbTxn) => {
+      const toPush = await dbTxn.db
+        .selectFrom('blob_push_event')
+        .selectAll()
+        .forUpdate()
+        .skipLocked()
+        .where('confirmedAt', 'is', null)
+        .execute()
```

Comment:
Two questions here:
 - there isn't an index supporting the `confirmedAt` checks, I don't think.  Maybe we need one as this table grows in size?
 - Is there intended to be a check on `attempts` here?

Reply:
good call on the index 👍 

and yup got that attempts check on the repo event poller but slipped by on the other too